rootston/layout.c

```diff
@@ -0,0 +1,233 @@
+	layout->wlr_layout = wlr_output_layout_create();
+	layout->current_config = config;
+
+	return layout;
+}
+
+void roots_layout_destroy(struct roots_layout *layout) {
+	if (layout) {
+		struct roots_layout_rule *rule, *tmp;
+		wl_list_for_each_safe(rule, tmp, &layout->rules, link) {
+			wl_list_remove(&rule->link);
+			free(rule);
+		}
+
+		free(layout);
+	}
+}
+
+static void apply_rule(struct roots_layout *layout,
+		struct roots_layout_rule *rule) {
+
+	rule->configured = false;
+
+	if (rule->config) {
+		if (rule->config->configuration !=
+				WLR_OUTPUT_LAYOUT_OUTPUT_CONFIGURATION_FIXED) {
+			struct roots_layout_rule *ref =
+				get_rule(layout, rule->config->reference_output);
+			if (ref) {
+				wlr_output_layout_add_relative(layout->wlr_layout,
+						rule->output->wlr_output, ref->output->wlr_output,
+						rule->config->configuration);
+				rule->configured = true;
+			} else {
+				wlr_output_layout_add(layout->wlr_layout,
+						rule->output->wlr_output, 0, 0);
+			}
+		} else {
+			wlr_output_layout_add(layout->wlr_layout, rule->output->wlr_output,
+					rule->config->x, rule->config->y);
+			rule->configured = true;
+		}
+	} else {
+		wlr_output_layout_add(layout->wlr_layout,
+				rule->output->wlr_output, 0, 0);
+	}
+
+	struct roots_layout_rule *ref_rule;
+	wl_list_for_each(ref_rule, &layout->rules, link) {
+		if (!ref_rule->configured && ref_rule->config) {
+			if (strcmp(rule->output->wlr_output->name,
+						ref_rule->config->reference_output) == 0) {
+				wlr_output_layout_add_relative(layout->wlr_layout,
+						ref_rule->output->wlr_output, rule->output->wlr_output,
+						ref_rule->config->configuration);
+				ref_rule->configured = true;
+			}
+		}
+	}
+}
+
+void roots_layout_add_output(struct roots_layout *layout,
+		struct roots_output *output) {
+
+	struct roots_layout_rule *new_rule;
+
+	wl_list_for_each(new_rule, &layout->rules, link) {
+		if (new_rule->output == output) {
+			return;
+		}
+	}
+
+	new_rule = calloc(1, sizeof(struct roots_layout_rule));
+
+	if (!new_rule) {
+		wlr_log(L_ERROR, "Could not allocate layout rule");
+		return;
+	}
+
+	new_rule->output = output;
+
+	struct roots_layout_rule_config *c_rule;
+	wl_list_for_each(c_rule, &layout->current_config->rules, link) {
+		if (strcmp(output->wlr_output->name, c_rule->output_name) == 0) {
+			new_rule->config = c_rule;
+			break;
+		}
+	}
+
+	wl_list_insert(&layout->rules, &new_rule->link);
+	apply_rule(layout, new_rule);
+}
+
+void roots_layout_remove_output(struct roots_layout *layout,
+		struct roots_output *output) {
+	struct roots_layout_rule *rule, *tmp;
+
+	wl_list_for_each_safe(rule, tmp, &layout->rules, link) {
+		if (rule->output == output) {
+			wl_list_remove(&rule->link);
+			free(rule);
+		} else if (rule->config && rule->config->reference_output &&
+				strcmp(output->wlr_output->name,
+					rule->config->reference_output) == 0) {
+			rule->configured = false;
+		}
+	}
+}
+
+void roots_layout_reflow(struct roots_layout *layout) {
```

Comment:
Sorry if this is obvious, but why is this required? I don't even understand what this functions does.

Reply:
It finds a position for the outputs that rootston can't configure (no configuration is available or the reference output is not connected) such that nog gaps accur in the layout. The first loop obtains the extents of the outputs that rootston was able to configure, the second loop moves the outputs to a proper position.

Reply:
You probably have already discussed about this, but why can't this happen in `wlr_output_layout`?

Reply:
The idea was to make the `auto` configuration not too complex and have it as a set it and forget type of configuration for examples and prototypes. Having a very specific behavior for these kind of outputs in `wlr_output_layout` might be to prescriptive for how compositors should handle outputs when they cannot be fully configured. Moving this into the `wlr_output_layout` would cater this specific case, but might not be sufficient for other compositors. As a minor point, it also shows how one can extend the `wlr_output_layout` and allowed me to see where it could be improved. https://github.com/swaywm/wlroots/pull/1002#issuecomment-392530341

Reply:
Can you add a comment explaining what this function does, since it's not obvious from its name?